### DISCUSSION TITLE
v10: TreeController, clarify sync/async methods

### DISCUSSION THREAD

**Discussion by JoseMarcenaro** [umbraco/Umbraco-CMS#12567] [Features and ideas]
There are two versions of the `GetTreeNodes` and `GetMenuForNode` methods.
![image](https://user-images.githubusercontent.com/923289/173084147-cbeb8d78-493e-486d-b260-214e16de77ab.png)

I assume implementing the async version is the recommended way to go - in fact, in v9 the sync methods were flagged as [Obsolete("Use the async version instead")].

In v10, the Obsolete attribute has been removed from the sync methods... and they both need to be implemented because the methods are abstract.

So my question is: what are the async methods used for? Are they even invoked? If so, why keep forcing the sync ones as abstract?

Thanks in advance,
Jose

**Comment by nzdev** [comment]
Ideally the sync methods would have been removed in v10 and the async ones be abstract instead of virtual, perhaps one for v11 now.
The async methods are the ones that umbraco calls, and their default implementation is the corresponding sync method, allowing for async code in tree controllers.
